### PR TITLE
runtime: add runtime loglevel config in the configuration.toml

### PR DESCRIPTION
**Pull request by flyflypeng**
reason: currently containerd-shim-kata-v2 binary has as an option
` -debug` to control print runtime log into the journal log.
However, runtime default log level is `warn`, which is too little
for debugging usage.
This patch adds a new config `loglevel` under runtime section  in
the configuration.toml to specifies the log level, which makes user
to select the log level simply.

Fixes: #366

Signed-off-by: flyflypeng <flyflyflypeng@gmail.com>

### CODE REVIEW

**Comment by fidencio** [comment]
@flyflypeng, the way Kata Containers works nowadays is by setting the `enable_debug` flag for the components you need, considering that `-debug` was passed to the `containerd-shim-kata-v2` binary, which has to be set in the CRI engine (CRI-O or containerd).

I, myself, am not against this change but from a quick look at the code it doesn't reflect into the agent code.  Or did I miss it?


I would really like to hear inputs from @amshinde, @bergwolf, @egernst, and @sameo

File: src/runtime/pkg/katautils/config.go [hunk 1]
```diff
@@ -1104,7 +1105,14 @@ func LoadConfiguration(configPath string, ignoreLogging bool) (resolvedConfigPat
...
 
 	config.Debug = tomlConf.Runtime.Debug
-	if !tomlConf.Runtime.Debug {
+	config.LogLevel, err = logrus.ParseLevel(tomlConf.Runtime.LogLevel)
+	if err != nil {
+		return "", oci.RuntimeConfig{}, err
+	}
+
+	if tomlConf.Runtime.Debug {
+		kataUtilsLogger.Logger.Level = config.LogLevel
```
Comment (jodh-intel) [inline src/runtime/pkg/katautils/config.go:1114]: This is neat as it doesn't break existing users but it's slightly confusing because you need to enable debug to change the level to something other than debug.

I wonder if instead we should just deprecate `enable_debug`. Something like:

```go
if tomlConf.Runtime.LogLevel != "" {
    kataUtilsLogger.Logger.Level = config.LogLevel
} else if tomlConf.Runtime.Debug {
     kataUtilsLogger.Logger.Level = Level.DebugLevel
}
```

If we do something like that shown above, we should also display a warning message similar to:

```
'enable_debug=' is deprecated - use 'loglevel=' instead"
```

... but note that actually outputting that message needs care since:

- This code is setting up the log levels and we always want that message to be displayed.
- Displaying the message "too early" will upset the container manager, which expects certain early output from the runtime and will fail if it sees something it doesn't expect.

We'd also need to think carefully about actually _removing_ `enable_debug=` since it would need changes to atleast:

- Docs.
- A bunch of scripts in the tests repo.

wdyt @amshinde, @bergwolf, @fidencio, @egernst?

**Review by jodh-intel — COMMENTED** [review]
Hi @flyflypeng - Thanks for raising. This is something we've been meaning to add for a long time now ;) I've added a comment but I agree with @fidencio that it would probably be best if we update the code for the runtime and the agent at the same time for consistency.

The code for the agent is here:

- https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/kata_agent.go#L312..L314

Basically, we need to pass `"agent.log=$level"` rather than hard-coding `"debug"`.

File: src/runtime/pkg/katautils/config.go (hunk 1 above)
Comment (amshinde) [reply src/runtime/pkg/katautils/config.go:1114]: @jodh-intel  +1 from me on removing enable_debug altogether although we may need to make quite a few changes in our logs and testing scripts. Having the ability to set the log level as added by this PR is of value to the user but having the enable_debug option is quite confusing and may conflict with the log level set by the option introduced in this PR.

Comment (flyflypeng) [reply src/runtime/pkg/katautils/config.go:1114]: @jodh-intel @fidencio I agree with replacing `enable_debug` config with `loglevel` config. If all committers reach an agreement with this change, I will modify the patch to support this change.

Comment (fidencio) [reply src/runtime/pkg/katautils/config.go:1114]: @jodh-intel, worries me quite a bit we remove an option and add a totally different one in between **minor** releases.

This may break compatibility for users who have their custom config / custom scripts to change those knobs in their nodes.

I sincerely think something as huge as this should be targeting a **major** release instead, being material for 3.0 rather than 2.x.

**Comment by fidencio** [comment]
I've added this PR for discussion in the next Architecture Committee meeting: https://etherpad.opendev.org/p/Kata_Containers_2022_Architecture_Committee_Mtgs#L23

**Comment by fidencio** [comment]
We've discussed this change as part of the architecture committee meeting and we think we should go ahead with this PR.
We should **not** break the current approach, but having the possibility to set the log_level right now is something desired.

**Comment by jodh-intel** [comment]
Looking forward to seeing this land @flyflypeng ! ;)

**Comment by flyflypeng** [comment]
> We've discussed this change as part of the architecture committee meeting and we think we should go ahead with this PR. We should **not** break the current approach, but having the possibility to set the log_level right now is something desired.

According to your reply,  I am still confused about how to implement this PR, keep the `enable_debug` config option and add a new `log_level` option, or others?

**Comment by flyflypeng** [comment]
> Looking forward to seeing this land @flyflypeng ! ;)

This may take some time， I am learning the Rust now which is a little hard for me.

**Comment by liubin** [comment]
@flyflypeng CI has some errors:

> tainers/src/github.com/kata-containers/kata-containers/src/runtime/pkg/katatestutils
INFO: Running golangci-lint on /home/runner/work/kata-containers/kata-containers/src/github.com/kata-containers/kata-containers/src/runtime/pkg/katautils
Error: config-settings.go:76:7: `defaultLogLevel` is unused (deadcode)
const defaultLogLevel string = "warn"
      ^
Error: config.go:143:14: fieldalignment: struct of size 160 could be 152 (govet)
type runtime struct {
             ^
make: *** [Makefile:40: static-checks] Error 1
Error: Process completed with exit code 2.

**Comment by fengwang666** [comment]
There is not any update for a few months. @flyflypeng Are you still able to finish it?

**Comment by fengwang666** [comment]
fyi, since @flyflypeng hasn't replied yet, I volunteer to take over the PR and finish it.

**Comment by yumingqiao** [comment]
Hi Kata community, I am working closely with @fengwang666. After coordinating with @fengwang666 , I volunteer to take over this change.

**Comment by fidencio** [comment]
Closing this one in favour of https://github.com/kata-containers/kata-containers/pull/6122
